packages/web/utils/background-routes/index.ts

```diff
@@ -0,0 +1,166 @@
+import {
+  OptimizedRoutesParams,
+  RouteWithInAmount,
+  SplitTokenInQuote,
+  Token,
+  TokenOutGivenInRouter,
+} from "@osmosis-labs/pools";
+
+import {
+  checkResponseAndDecodeError,
+  decodeRouteWithInAmount,
+  decodeSplitTokenInQuote,
+  encodeCalculateTokenOutByTokenInParameters,
+  encodeGetOptimizedRoutesByTokenInParameters,
+  encodeOptimizedRoutesParams,
+  encodeRouteByTokenInParameters,
+} from "./coding";
+import {
+  EncodedRequest,
+  EncodedResponse,
+  OptimizedRoutesWorker,
+} from "./worker";
+
+/** Router that delegates search problem to background thread, useful for unblocking the main thread for UI/DOM updates. */
+export class BackgroundRoutes implements TokenOutGivenInRouter {
```

Comment:
I think this should work fine, but I wonder if the use of serial numbers and collecting pending responses is overcomplicating things? We aren't making any server connections here right? Typically, web workers are just passing messages back and forth https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Using_web_workers

Wondering what the impetus was behind using this model? Am I missing something?

Reply:
I use that to maintain the Promise-based API in the router interface.

We use the serial number to identify the message as it's passed to the worker, then we use the same serial number in the future to identify the right response in the response pool and resolve the promise with that response. I don't believe there's any guarantee of order as response messages are received from worker. The use of the loop and timeout isn't ideal, but it seems to work fine for the most part. It is a bit fragile.

I found this from Google: [Comlink](https://github.com/GoogleChromeLabs/comlink). Intends to modernize the API around web workers as a Promise-based RPC API, but IMO what we have works fine as we could avoid pulling in that dependency (which hasn't been updated in a while).

Reply:
Also touched on this in top level README.